### ISSUE TITLE
Ill-conditioned gradients in QR

### ISSUE THREAD

**Issue by Lezcano** [pytorch/pytorch#42792]
## 🐛 Bug

The gradients in the QR decomposition, as it is implemented without pivoting, they are __very__ ill-conditioned around matrices that are not full-rank.

## To Reproduce

```python
import torch

def pos_qr(A):
    """ Returns the QR decomposition with R.diag() > 0 """
    Q, R = torch.qr(A)
    d = torch.diag_embed(torch.diag(R).sign())
    # make fast
    return Q @ d, d @ R


def same_function_different_grad(A, G, conj):
    """ Compute one of: 
       - torch.qr(A).Q
       - U.t() @ torch.qr(U @ A).Q
    for an orthogonal U. 
    In both cases the result should be the same
    """
    if conj:
        U = torch.qr(torch.rand(8, 8)).Q
        A = U @ A
    Q, _ = pos_qr(A)
    if conj:
        Q = U.t() @ Q
    out = torch.autograd.grad(Q, A, G)[0]
    return Q, out

# A is not full-rank by construction
A = torch.cat([torch.rand(8, 3, requires_grad=True), torch.zeros(8, 5, requires_grad=True)], dim=1)
# Some fixed gradient
G = torch.rand(8, 8)
Q1, B1 = same_function_different_grad(A, G, False)
Q2, B2 = same_function_different_grad(A, G, True)
print(torch.norm(Q1 - Q2))
print(torch.norm(B1 - B2))
```

Example output:
```
tensor(6.3826e-07, grad_fn=<NormBackward0>)                                                                            
tensor(2.1728)
```

## Expected behavior

This is expected due to the current implementation of the QR decomposition (without pivoting) but it can very easily be prevented if one performs QR with pivoting. Note that this is already implemented in BLAS and MAGMA (cf. #10454), so it should not be too difficult to add to PyTorch.

A side-problem derived from this is that, since the gradients with respect to the Q in the QR decomposition are not well-defined for matrices that are not full-rank, the gradient in the QR decomposition is not currently tested in PyTorch, as the testing framework does not allow for non-deterministic gradients.

@nikitaved 

cc @ezyang @SsnL @albanD @zou3519 @gqchen @vincentqb @vishwakftw @jianyuh

**Comment by ezyang** [comment]
If anyone at QS wants to pick this up, I think this would be pretty good!

**Comment by nikitaved** [comment]
Then I am taking this one.

**Comment by nikitaved** [comment]
I did check the gradients, it involves solving systems of overdetermined systems of linear equations, for example for `R^{-T}`. Now, the solution is derived using `triangular_solve`. I do not know how it handles zero rows (`R` must have zero rows because of rank deficiency). If this handing is poor, we will also need to modify the backward formula.

**Comment by nikitaved** [comment]
Looks like `triangular_solve` does handle rank-deficiency but only if in `Ax = b` `b in range(A)`. Otherwise have a look at this:
```python
In [22]: a
Out[22]: 
tensor([[1., 0., 0.],
        [0., 1., 0.],
        [0., 0., 0.]])

In [23]: b
Out[23]: 
tensor([[0.9509],
        [0.8624],
        [0.8157]])

In [24]: torch.triangular_solve(b, a)
Out[24]: 
torch.return_types.triangular_solve(
solution=tensor([[0.9509],
        [0.8624],
        [0.8157]]),
cloned_coefficient=tensor([[1., 0., 0.],
        [0., 1., 0.],
        [0., 0., 0.]]))


```
This system does not have a solution, however.

**Comment by nikitaved** [comment]
Now, when it comes to QR:
```python
In [31]: x = torch.cat([torch.rand(5, 3), torch.zeros(5, 2)], dim=1)

In [32]: x
Out[32]: 
tensor([[0.0280, 0.6064, 0.3813, 0.0000, 0.0000],
        [0.1165, 0.0986, 0.0408, 0.0000, 0.0000],
        [0.1970, 0.3721, 0.9981, 0.0000, 0.0000],
        [0.3110, 0.8561, 0.9606, 0.0000, 0.0000],
        [0.5922, 0.0450, 0.4453, 0.0000, 0.0000]])

In [33]: q, r = torch.qr(x)

In [34]: q
Out[34]: 
tensor([[-0.0395,  0.6028,  0.3252, -0.7207, -0.0993],
        [-0.1647,  0.0069,  0.2580,  0.2575, -0.9165],
        [-0.2785,  0.2236, -0.8866, -0.1642, -0.2440],
        [-0.4395,  0.6302,  0.1164,  0.5659,  0.2755],
        [-0.8370, -0.4352,  0.1677, -0.2591,  0.1215]])

In [35]: r
Out[35]: 
tensor([[-0.7075, -0.5578, -1.0946,  0.0000,  0.0000],
        [ 0.0000,  0.9694,  0.8649,  0.0000,  0.0000],
        [ 0.0000,  0.0000, -0.5638,  0.0000,  0.0000],
        [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000],
        [ 0.0000,  0.0000,  0.0000,  0.0000,  0.0000]])

```
So, it looks like QR does handle rank-deficiency. So, probably, it is the grad that requires inspection.

**Comment by nikitaved** [comment]
It seems I was able to detect the issue. Consider for simplicity that the input matrix `A` is square, so `QQ^T = I`.
Now, in the backward formula one of the intermediate steps involves solving the system
```
RX = (Q_grad -  QQ^T [= I]).
```
It is being solved with `triangular_solve`.
Now, the rhs is super likely to be full rank, while the rhs is not full-rank, hence there is no solution in such a case.
Another words, backward is incorrect ;(

**Comment by nikitaved** [comment]
By the way, the QR decomposition with pivoting is still great to have! And now I have a question, shall it be a separate function, or adding a flag to the existing frontend would suffice?
The version with pivoting returns 3 matrices, that is Q, R and a perturbation matrix P so that AP=QR, the diagonal values of R are sorted in a non-decreasing manner based on their module.

Current QR decomposition does not guaranteed, unfortunately, that rank(R[:, :k]) = k, which means that it is not just a simple matter of indexing to extract linearly independent rows of R/columns of Q which actually generate A.
This makes qr_backward even harder to implement in a rank-deficient case.

**Comment by vishwakftw** [comment]
> And now I have a question, shall it be a separate function, or adding a flag to the existing frontend would suffice?

It would be preferable to have it in the same function, perhaps with the flag `pivoting`.

**Comment by Lezcano** [comment]
This is expected behaviour and is now documented in `torch.linalg.qr`.
